File: scripts/setup-supabase.mjs
#!/usr/bin/env node
/**
 * Setup remoto Supabase para web-reservas.
 *
 * Requiere en .env.local:
 *   NEXT_PUBLIC_SUPABASE_URL
 *   NEXT_PUBLIC_SUPABASE_ANON_KEY
 *   SUPABASE_SERVICE_ROLE_KEY
 *
 * Y UNA de:
 *   SUPABASE_DB_PASSWORD  (Settings → Database → Database password)
 *   DATABASE_URL          (postgres connection string)
 *   SUPABASE_ACCESS_TOKEN (token personal sbp_... del dueño del proyecto)
 *
 * Uso: node scripts/setup-supabase.mjs
 */
import { readFileSync, existsSync, readdirSync } from "fs";
import { resolve, dirname, join } from "path";
import { fileURLToPath } from "url";
import { createClient } from "@supabase/supabase-js";
import pg from "pg";

const __dirname = dirname(fileURLToPath(import.meta.url));
const root = resolve(__dirname, "..");

function loadEnv() {
  const env = { ...process.env };
  for (const name of [".env.local", ".env"]) {
    const p = join(root, name);
    if (!existsSync(p)) continue;
    for (const line of readFileSync(p, "utf8").split("\n")) {
      const m = line.match(/^([A-Z0-9_]+)=(.*)$/);
      if (!m) continue;
      env[m[1]] = m[2].replace(/^["']|["']$/g, "");
    }
  }
  return env;
}

const env = loadEnv();
const url = env.NEXT_PUBLIC_SUPABASE_URL;
const anon = env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
const service = env.SUPABASE_SERVICE_ROLE_KEY;
const ref = url?.match(/https:\/\/([a-z0-9]+)\.supabase\.co/)?.[1];

if (!url || !anon || !service || !ref) {
  console.error("Faltan NEXT_PUBLIC_SUPABASE_URL / ANON / SERVICE_ROLE en .env.local");
  process.exit(1);
}

const sqlPath = join(root, "supabase/migrations/0001_init.sql");
const sql = readFileSync(sqlPath, "utf8");

async function runSqlViaPg(connectionString) {
  const client = new pg.Client({
    connectionString,
    ssl: { rejectUnauthorized: false },
  });
  await client.connect();
  try {
    await client.query(sql);
    console.log("✓ SQL aplicado via Postgres");
  } finally {
    await client.end();
  }
}

async function runSqlViaManagement(token) {
  const endpoint = `https://api.supabase.com/v1/projects/${ref}/database/query`;
  const res = await fetch(endpoint, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ query: sql }),
  });
  const text = await res.text();
  if (!res.ok) {
    throw new Error(`Management API ${res.status}: ${text.slice(0, 400)}`);
  }
  console.log("✓ SQL aplicado via Management API");
}

function dbUrlFromPassword(password) {
  const enc = encodeURIComponent(password);
  // pooler session mode + direct
  return [
    `postgresql://postgres.${ref}:${enc}@aws-0-us-east-1.pooler.supabase.com:5432/postgres`,
    `postgresql://postgres:${enc}@db.${ref}.supabase.co:5432/postgres`,
  ];
}

async function applySql() {
  if (env.DATABASE_URL) {
    await runSqlViaPg(env.DATABASE_URL);
    return;
  }
  if (env.SUPABASE_DB_PASSWORD) {
    const urls = dbUrlFromPassword(env.SUPABASE_DB_PASSWORD);
    let lastErr;
    for (const u of urls) {
      try {
        await runSqlViaPg(u);
        return;
      } catch (e) {
        lastErr = e;
        console.warn("PG intento falló:", e.message);
      }
    }
    throw lastErr;
  }
  if (env.SUPABASE_ACCESS_TOKEN) {
    await runSqlViaManagement(env.SUPABASE_ACCESS_TOKEN);
    return;
  }
  throw new Error(
    "Necesito SUPABASE_DB_PASSWORD o DATABASE_URL o SUPABASE_ACCESS_TOKEN para crear tablas"
  );
}

async function uploadPhotos() {
  const sb = createClient(url, service, {
    auth: { persistSession: false, autoRefreshToken: false },
  });
  const photosDir = join(root, "public/photos");
  const files = [
    { file: "depto-hero.jpg", path: "gallery/depto-hero.jpg", alt: "Hero", cover: true, sort: 0 },
    { file: "depto-01.jpg", path: "gallery/depto-01.jpg", alt: "Living", cover: false, sort: 1 },
    { file: "depto-02.jpg", path: "gallery/depto-02.jpg", alt: "Ambiente", cover: false, sort: 2 },
    { file: "depto-03.jpg", path: "gallery/depto-03.jpg", alt: "Detalle", cover: false, sort: 3 },
    { file: "depto-04.jpg", path: "gallery/depto-04.jpg", alt: "Vista", cover: false, sort: 4 },
  ];

  // clean previous gallery rows pointing to local paths
  await sb.from("photos").delete().like("storage_path", "/%");

  for (const f of files) {
    const local = join(photosDir, f.file);
    if (!existsSync(local)) {
      console.warn("skip missing", f.file);
      continue;
    }
    const buf = readFileSync(local);
    const { error: upErr } = await sb.storage
      .from("property-photos")
      .upload(f.path, buf, { contentType: "image/jpeg", upsert: true });
    if (upErr) throw new Error(`upload ${f.file}: ${upErr.message}`);

    const { data: existing } = await sb
      .from("photos")
      .select("id")
      .eq("storage_path", f.path)
      .maybeSingle();

    if (existing) {
      await sb
        .from("photos")
        .update({ alt: f.alt, sort_order: f.sort, is_cover: f.cover })
        .eq("id", existing.id);
    } else {
      const { error } = await sb.from("photos").insert({
        storage_path: f.path,
        alt: f.alt,
        sort_order: f.sort,
        is_cover: f.cover,
      });
      if (error) throw new Error(`photos insert ${f.file}: ${error.message}`);
    }
    console.log("✓ foto", f.file);
  }

  const { data: prop } = await sb.from("property").select("id").limit(1).maybeSingle();
  if (prop) {
    await sb
      .from("property")
      .update({ cover_photo_path: "gallery/depto-hero.jpg" })
      .eq("id", prop.id);
    console.log("✓ cover_photo_path set");
  }
}

async function verify() {
  const sb = createClient(url, anon);
  const { data: prop, error: e1 } = await sb.from("property").select("name,whatsapp_e164").limit(1).maybeSingle();
  if (e1) throw e1;
  const { count: avail } = await sb
    .from("availability")
    .select("*", { count: "exact", head: true })
    .eq("status", "available");
  const { count: photos } = await sb.from("photos").select("*", { count: "exact", head: true });
  console.log("✓ verify property:", prop);
  console.log("✓ available nights:", avail);
  console.log("✓ photos:", photos);
}

async function main() {
  console.log("Proyecto", ref);
  await applySql();
  await uploadPhotos();
  await verify();
  console.log("\nListo. Reiniciá el dev server para cargar .env.local.");
}

main().catch((e) => {
  console.error("\nERROR:", e.message || e);
  process.exit(1);
});

File: scripts/setup_supabase.py
#!/usr/bin/env python3
"""Setup Supabase remoto para web-reservas (schema + storage + fotos).

Requiere .env.local con URL/anon/service_role y UNA de:
  SUPABASE_DB_PASSWORD
  DATABASE_URL
  SUPABASE_ACCESS_TOKEN (sbp_...)

Uso:
  python3 scripts/setup_supabase.py
"""
from __future__ import annotations

import json
import os
import re
import sys
import urllib.error
import urllib.request
from pathlib import Path

try:
    import psycopg2
except ImportError:
    print("Falta psycopg2", file=sys.stderr)
    sys.exit(1)

ROOT = Path(__file__).resolve().parents[1]
MIGRATIONS_DIR = ROOT / "supabase/migrations"
PHOTOS_DIR = ROOT / "public/photos"


def load_env() -> dict[str, str]:
    env = dict(os.environ)
    for name in (".env.local", ".env"):
        p = ROOT / name
        if not p.exists():
            continue
        for line in p.read_text().splitlines():
            m = re.match(r"^([A-Z0-9_]+)=(.*)$", line.strip())
            if not m:
                continue
            env[m.group(1)] = m.group(2).strip().strip('"').strip("'")
    return env


def project_ref(url: str) -> str:
    m = re.match(r"https://([a-z0-9]+)\.supabase\.co", url)
    if not m:
        raise SystemExit(f"URL inválida: {url}")
    return m.group(1)


def apply_sql_pg(conninfo: str, sql: str) -> None:
    conn = psycopg2.connect(conninfo, sslmode="require")
    conn.autocommit = True
    try:
        with conn.cursor() as cur:
            cur.execute(sql)
        print("✓ SQL aplicado via Postgres")
    finally:
        conn.close()


def apply_sql_management(token: str, ref: str, sql: str) -> None:
    endpoint = f"https://api.supabase.com/v1/projects/{ref}/database/query"
    req = urllib.request.Request(
        endpoint,
        data=json.dumps({"query": sql}).encode(),
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=120) as res:
            res.read()
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", "replace")
        raise SystemExit(f"Management API {e.code}: {body[:500]}") from e
    print("✓ SQL aplicado via Management API")


def candidate_db_urls(ref: str, password: str) -> list[str]:
    from urllib.parse import quote_plus

    enc = quote_plus(password)
    # regions comunes + direct host
    regions = [
        "us-east-1",
        "us-east-2",
        "us-west-1",
        "us-west-2",
        "sa-east-1",
        "eu-west-1",
        "eu-west-2",
        "eu-central-1",
    ]
    urls = [f"postgresql://postgres:{enc}@db.{ref}.supabase.co:5432/postgres"]
    for r in regions:
        urls.append(
            f"postgresql://postgres.{ref}:{enc}@aws-0-{r}.pooler.supabase.com:5432/postgres"
        )
        urls.append(
            f"postgresql://postgres.{ref}:{enc}@aws-1-{r}.pooler.supabase.com:5432/postgres"
        )
    return urls


def rest(url: str, key: str, method: str, path: str, body: bytes | None = None, extra: dict | None = None):
    headers = {
        "apikey": key,
        "Authorization": f"Bearer {key}",
    }
    if extra:
        headers.update(extra)
    req = urllib.request.Request(url.rstrip("/") + path, data=body, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=60) as res:
            return res.status, res.read()
    except urllib.error.HTTPError as e:
        return e.code, e.read()


def upload_photos(base: str, service: str) -> None:
    files = [
        ("depto-hero.jpg", "gallery/depto-hero.jpg", "Hero", True, 0),
        ("depto-01.jpg", "gallery/depto-01.jpg", "Living", False, 1),
        ("depto-02.jpg", "gallery/depto-02.jpg", "Ambiente", False, 2),
        ("depto-03.jpg", "gallery/depto-03.jpg", "Detalle", False, 3),
        ("depto-04.jpg", "gallery/depto-04.jpg", "Vista", False, 4),
    ]

    # wipe local-path leftovers if any
    rest(
        base,
        service,
        "DELETE",
        "/rest/v1/photos?storage_path=like./%",
        extra={"Prefer": "return=minimal"},
    )

    for local_name, storage_path, alt, is_cover, sort in files:
        local = PHOTOS_DIR / local_name
        if not local.exists():
            print("skip missing", local_name)
            continue
        data = local.read_bytes()
        status, body = rest(
            base,
            service,
            "POST",
            f"/storage/v1/object/property-photos/{storage_path}",
            body=data,
            extra={
                "Content-Type": "image/jpeg",
                "x-upsert": "true",
            },
        )
        if status not in (200, 201):
            # try update path
            status2, body2 = rest(
                base,
                service,
                "PUT",
                f"/storage/v1/object/property-photos/{storage_path}",
                body=data,
                extra={"Content-Type": "image/jpeg"},
            )
            if status2 not in (200, 201):
                raise SystemExit(
                    f"upload {local_name} failed {status}/{status2}: {body[:200]!r} {body2[:200]!r}"
                )

        # upsert photo row: delete+insert simple
        rest(
            base,
            service,
            "DELETE",
            f"/rest/v1/photos?storage_path=eq.{storage_path}",
            extra={"Prefer": "return=minimal"},
        )
        payload = json.dumps(
            {
                "storage_path": storage_path,
                "alt": alt,
                "sort_order": sort,
                "is_cover": is_cover,
            }
        ).encode()
        st, b = rest(
            base,
            service,
            "POST",
            "/rest/v1/photos",
            body=payload,
            extra={"Content-Type": "application/json", "Prefer": "return=minimal"},
        )
        if st not in (200, 201):
            raise SystemExit(f"photos insert {local_name}: {st} {b[:300]!r}")
        print("✓ foto", local_name)

    st, b = rest(base, service, "GET", "/rest/v1/property?select=id&limit=1")
    if st != 200:
        raise SystemExit(f"property get: {st} {b[:200]!r}")
    rows = json.loads(b.decode())
    if rows:
        pid = rows[0]["id"]
        payload = json.dumps({"cover_photo_path": "gallery/depto-hero.jpg"}).encode()
        st, b = rest(
            base,
            service,
            "PATCH",
            f"/rest/v1/property?id=eq.{pid}",
            body=payload,
            extra={"Content-Type": "application/json", "Prefer": "return=minimal"},
        )
        if st not in (200, 204):
            raise SystemExit(f"cover update: {st} {b[:200]!r}")
        print("✓ cover_photo_path")


def verify(base: str, anon: str) -> None:
    st, b = rest(base, anon, "GET", "/rest/v1/property?select=name,whatsapp_e164&limit=1")
    print("property", st, b.decode()[:300])
    st, b = rest(
        base,
        anon,
        "GET",
        "/rest/v1/availability?status=eq.available&select=night_date&limit=5",
    )
    print("availability sample", st, b.decode()[:300])
    st, b = rest(base, anon, "GET", "/rest/v1/photos?select=storage_path,is_cover&order=sort_order")
    print("photos", st, b.decode()[:400])


def main() -> None:
    env = load_env()
    base = env.get("NEXT_PUBLIC_SUPABASE_URL", "")
    anon = env.get("NEXT_PUBLIC_SUPABASE_ANON_KEY", "")
    service = env.get("SUPABASE_SERVICE_ROLE_KEY", "")
    if not base or not anon or not service:
        raise SystemExit("Faltan keys en .env.local")
    ref = project_ref(base)
    migration_files = sorted(MIGRATIONS_DIR.glob("*.sql"))
    if not migration_files:
        raise SystemExit(f"No hay migrations en {MIGRATIONS_DIR}")
    print("Proyecto", ref)
    print("Migrations:", ", ".join(p.name for p in migration_files))

    def apply_all(apply_fn) -> None:
        for path in migration_files:
            print("→", path.name)
            apply_fn(path.read_text())

    if env.get("DATABASE_URL"):
        apply_all(lambda sql: apply_sql_pg(env["DATABASE_URL"], sql))
    elif env.get("SUPABASE_ACCESS_TOKEN"):
        apply_all(
            lambda sql: apply_sql_management(env["SUPABASE_ACCESS_TOKEN"], ref, sql)
        )
    elif env.get("SUPABASE_DB_PASSWORD"):
        last = None
        applied = False
        for u in candidate_db_urls(ref, env["SUPABASE_DB_PASSWORD"]):
            try:
                apply_all(lambda sql, url=u: apply_sql_pg(url, sql))
                last = None
                applied = True
                break
            except Exception as e:  # noqa: BLE001
                last = e
                print("intento PG falló:", str(e)[:120])
        if not applied:
            raise SystemExit(f"No pude conectar a Postgres: {last}")
    else:
        raise SystemExit(
            "Necesito SUPABASE_DB_PASSWORD o DATABASE_URL o SUPABASE_ACCESS_TOKEN"
        )

    upload_photos(base, service)
    verify(base, anon)
    print("\nListo. Reiniciá `npm run dev` para cargar .env.local.")


if __name__ == "__main__":
    main()
